packages/react-devtools-shared/src/devtools/views/Components/TreeContext.js

```diff
@@ -372,6 +380,92 @@ function reduceTreeState(store: Store, state: State, action: Action): State {
+        if (store.errorsAndWarnings.size === 0) {
+          return state;
+        }
+
+        // Mapping of ID -> Index in the flat tree
+        const elementIndicesWithErrorsOrWarnings: [
+          number,
+          number,
+        ][] = Array.from(store.errorsAndWarnings.keys())
+          .map(elementId => {
+            // $FlowFixMe https://github.com/facebook/flow/issues/1414
+            return [elementId, store.getIndexOfElementID(elementId)];
+          })
+          .filter(([, elementIndex]) => {
+            return elementIndex !== null;
+          });
+        elementIndicesWithErrorsOrWarnings.sort(([, a], [, b]) => {
+          return b - a;
+        });
+
+        const flatSelectedElementIndex =
+          // Selected element in a collapsed tree?
+          selectedElementIndex === null && selectedElementID !== null
+            ? store.getIndexOfElementID(selectedElementID)
+            : selectedElementIndex;
+        const predecessors = elementIndicesWithErrorsOrWarnings.filter(
+          ([, elementIndex]) => {
+            return (
+              flatSelectedElementIndex === null ||
+              elementIndex < flatSelectedElementIndex
+            );
+          },
+        );
+        if (predecessors.length === 0) {
+          selectedElementID = elementIndicesWithErrorsOrWarnings[0][0];
+        } else {
+          selectedElementID = predecessors[0][0];
+        }
+        selectedElementIndex = store.getIndexOfElementID(selectedElementID);
+        lookupIDForIndex = false;
+        break;
+      }
+      case 'SELECT_NEXT_ELEMENT_WITH_ERROR_OR_WARNING_IN_TREE': {
+        if (store.errorsAndWarnings.size === 0) {
+          return state;
+        }
+
+        // Mapping of ID -> Index in the flat tree
+        const elementIndicesWithErrorsOrWarnings: [
+          number,
+          number,
+        ][] = Array.from(store.errorsAndWarnings.keys())
+          .map(elementId => {
+            // $FlowFixMe https://github.com/facebook/flow/issues/1414
+            return [elementId, store.getIndexOfElementID(elementId)];
+          })
+          .filter(([, elementIndex]) => {
+            return elementIndex !== null;
+          });
+        elementIndicesWithErrorsOrWarnings.sort(([, a], [, b]) => {
+          return a - b;
+        });
+
+        const flatSelectedElementIndex =
+          // Selected element in a collapsed tree?
+          selectedElementIndex === null && selectedElementID !== null
+            ? store.getIndexOfElementID(selectedElementID)
+            : selectedElementIndex;
+        const successors = elementIndicesWithErrorsOrWarnings.filter(
+          ([, elementIndex]) => {
+            return (
+              flatSelectedElementIndex === null ||
+              elementIndex > flatSelectedElementIndex
+            );
+          },
+        );
+        if (successors.length === 0) {
+          selectedElementID = elementIndicesWithErrorsOrWarnings[0][0];
```

Comment:
This line still seems problematic. I don't know how to break it in a test (yet) but it breaks my extension when I try using it on Facebook.

Reply:
We return early in https://github.com/facebook/react/blob/64ad2e25324016e94167d6d1d29f07e362fe221b/packages%2Freact-devtools-shared%2Fsrc%2Fdevtools%2Fviews%2FComponents%2FTreeContext.js#L427 if there are no errors. 

But we're not guarding against the case where we do have recorded errors but `store.getIndexOfElementID(elementId)` returns `null` for these fiberIDs. 

So `store.errorsAndWarnings.size` can be greater than 0 but `elementIndicesWithErrorsOrWarnings.length` can still be `0`.

Reply:
Right. That case _shouldn't_ occur but clearly it does so we need to track it down.

Reply:
Looks like we're not properly removing errors when fibers unmount in certain cases, which leaves the total number of errors in the map > 0. Once there are no more actually mounted errors, then this line throws.

Reply:
Oh I think I caught it (or at least one variant of it): error boundaries!

Reply:
Got a test

Reply:
Working on it.

Reply:
This test fails in two ways:
```js
it('from react during unmount get counted', () => {
  class ErrorBoundary extends React.Component {
    state = {error: null};
    static getDerivedStateFromError(error) {
      return {error};
    }
    render() {
      console.log('<ErrorBoundary>', !this.state.error);
      if (this.state.error) {
        return null;
      }
      return this.props.children;
    }
  }

  class BadRender extends React.Component {
    render() {
      console.error('test-only: I am about to throw!');
      throw new Error('test-only: Oops!');
    }
  }

  const container = document.createElement('div');
  withErrorsOrWarningsIgnored(
    ['test-only:', 'React will try to recreate this component tree'],
    () => {
      utils.act(() =>
        ReactDOM.render(
          <ErrorBoundary>
            <BadRender />
          </ErrorBoundary>,
          container,
        ),
      );
    },
  );
  expect(store).toMatchInlineSnapshot(`
    [root] ✕ 2, ⚠ 0
        <ErrorBoundary> ✕
  `);

  utils.act(() => TestRenderer.create(<Contexts />));

  utils.act(() => ReactDOM.unmountComponentAtNode(container));
  expect(state).toMatchInlineSnapshot(``);

  // Should be a noop
  selectNextErrorOrWarning();
  expect(state).toMatchInlineSnapshot();
});
```

First the snapshot fails with 4 errors instead of the two selected ones.

Second the `selectNextErrorOrWarning` call fails with the error I left a comment about above.

Pretty sure this is because we aren't cleaning up after fibers that get added (but not committed).

Reply:
Test: [`bc788d8` (#20463)](https://github.com/facebook/react/pull/20463/commits/bc788d82066ece29b858a072c8da262babf78202)

It comes back to

> Slightly related: Have we thought about what we want to do with errors of renders that are thrown away. For example, we render+error, suspend, render+error, commit. I guess right now we're logging both errors?

-- https://github.com/facebook/react/pull/20463#discussion_r545247815

There's no fiber we can scroll to (yet) so how do we treat errors from those? Since we can't look at them in the frontend we should probably just ignore them. The console will still include them though so it might be confusing?

Reply:
I think I have a fix.

Reply:
Generally speaking: Do we even want to display these warnings and if so:  how do we display them? Because right now it's a bit confusing that we count them but not show where it happened or what the message was.

Reply:
I'd like a little more time to think about this case.

Reply:
I have several new tests, and a fix (I think) but I'm trying to think through how to test various Suspense and error boundary cases.

Reply:
Check out commit 22c3f4b?

Reply:
FWIW this does seem to have fixed the error I saw on Facebook. (Just repro'ed and no longer breaks.)

Reply:
I think there's another error though. I don't see a runtime error but I do see invalid counts. :(

Reply:
I think I'm going to step away for a bit. Clear my head. Let me know if you hve any ideas!

Reply:
GitHub has given up on automatically updating this page for me. I Need to refresh before I write something so that it's not outdated by the time I post it.

> Check out commit 22c3f4b?

Looks good to me :+1: 

> I think there's another error though. I don't see a runtime error but I do see invalid counts. :(

I'm going to add a test for errors from forwardRef but other than that I don't think I'm able to help you there without a test. I don't have that many Suspense-heavy apps or apps using concurrent mode that I can test the extension on so I probably won't encounter as many stress cases as you might.

Reply:
Yeah fair enough.

One thing I'm thinking about removing is the separate timeout+flush for errors. I think we may be making things more difficult on ourselves to detect not-yet-suspended-but-still-waiting type states with the way things currently are. I think it would reduce potential edge cases if we just piggy backed on the next commit.

The only exception would be the clear methods– those would just insta-flush.

Reply:
> I don't have that many Suspense-heavy apps or apps using concurrent mode that I can test the extension on so I probably won't encounter as many stress cases as you might.

I don't _think_ it requires APIs like Suspense to repro for what it's worth. I was just able to repro the remaining problem (sometimes errors get left over in DevTools even after being unmounted– and you'll see a _No element found with id "..."_ message in the second level console) using one of our legacy apps that doesn't make use of any of the new APIs.

There's probably some other unmount related case that I'm not handling properly yet. I'm really sleepy so I don't trust myself to think about the edge cases right now. :)

Reply:
> I'm going to add a test for errors from forwardRef 

Hm, I was thinking about something else. I wanted to see what happens if the forwardRef wrapper errors during render as well as the component it wraps. But forwardRef doesn't have any errors/warnings during render (passing a wrong ref type throws, passing a function that doesn't handle the second argument warns when creating the element type not during render).

Going to use the latest build from now on and see if something fails. facebook.com doesn't log any warnings or errors for me that I can debug :wink:

Reply:
Yeah you'd have to be running a DEV build of Facebook to get any errors :)

But if you have other DEV sites you can use...

Reply:
So I'm debugging a case of this lingering "bug" right now– the UI shows a warning still, but the next/prev arrows do nothing (and so far as I can tell, there is no warning). I got a handle on the `Store` instance and its `store.errorsAndWarnings` map is also empty.

I think this is because the subscriptions are still listening for "errorsAndWarnings", which doesn't always get emitted (an we don't have tests for it specifically).

Reply:
I think I see a place or two misisng.

Reply:
I think I found the fix but I want to add a failing test first. Give me 5.

Reply:
Maybe just pushed a fix!

Reply:
YESSS seems like it fixed both places!

Reply:
> But if you have other DEV sites you can use...

Tried to upgrade a library in an app that uses an experimental React build and I did encounter some crashes. Will try to repro them over the weekend.

Reply:
Bummer! ok, keep me posted!

Reply:
I probably had an older version installed locally. The contained a bunch of Suspense fallbacks that had errors. 

After updating the extension the errors got removed from the devtools UI once the Suspense fallbacks got unmounted.

Reply:
Oh! Yay!

Reply:
I think my latest commit addresses the thing I was kind of worried about with the [0][0]